target_chains/ethereum/sdk/solidity/PriceLibrary.sol

```diff
@@ -0,0 +1,481 @@
+ */
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./PythStructs.sol";
+
+
+// Constants defining precision and limits for prices
+int32 constant PD_EXPO = -9;
+uint64 constant PD_SCALE = 1_000_000_000;
+uint64 constant MAX_PD_V_U64 = (1 << 28) - 1;
+
+/**
+ * @notice Functions for manipulating PythStructs.Price data types.
+ */
+library PriceLibrary {
+
+    /**
+     * @notice Get the price in a specified quote currency.
+     * @param self The base price.
+     * @param quote The quote price.
+     * @param resultExpo The exponent to scale the result to.
+     * @return Scaled Price struct.
+     */
+    function getPriceInQuote(
+        PythStructs.Price memory self,
+        PythStructs.Price memory quote,
+        int32 resultExpo
+    ) public pure returns (PythStructs.Price memory) {
+        // Divide the price self by the quote price and scale to desired exponent
+        PythStructs.Price memory divResult = div(self, quote);
+        return scaleToExponent(divResult, resultExpo);
+    }
+
+    /**
+     * @notice Calculate collateral valuation price based on deposits and rate discounts.
+     * @param self The base price.
+     * @param deposits Amount of deposits.
+     * @param depositsEndpoint Upper limit of deposits.
+     * @param rateDiscountInitial Initial rate discount.
+     * @param rateDiscountFinal Final rate discount.
+     * @param discountExponent Exponent for rate discounts.
+     * @return Adjusted Price struct.
+     */
+    function getCollateralValuationPrice(
+        PythStructs.Price memory self,
+        uint64 deposits,
+        uint64 depositsEndpoint,
+        uint64 rateDiscountInitial,
+        uint64 rateDiscountFinal,
+        int32 discountExponent
+    ) public pure returns (PythStructs.Price memory) {
+        // If the initial discount rate is greater than the final rate, return zero price
+        if (rateDiscountInitial < rateDiscountFinal) {
+            return PythStructs.Price(0, 0, 0, 0);
+        }
+
+        // Create Price structs for initial and final discount percentages
+        PythStructs.Price memory initialPercentage = PythStructs.Price({
+            price: int64(rateDiscountInitial),
+            conf: 0,
+            expo: discountExponent,
+            publishTime: 0
+        });
+        PythStructs.Price memory finalPercentage = PythStructs.Price({
+            price: int64(rateDiscountFinal),
+            conf: 0,
+            expo: discountExponent,
+            publishTime: 0
+        });
+
+        // Interpolate discount percentage based on deposits amount
+        PythStructs.Price memory discountInterpolated = affineCombination(
+            0,
+            initialPercentage,
+            int64(depositsEndpoint),
+            finalPercentage,
+            int64(deposits),
+            -9
+        );
+        // If interpolated price is zero, return zero price
+        if (discountInterpolated.price == 0) {
```

Comment:
would not do this. `confidence` could still be nonzero

Reply:
Sure updated it.

Reply:
> would not do this. `confidence` could still be nonzero

Sure updated it.